* TOC
  {:toc}

# WELCOME

Reliable detection of somatic variations is of critical importance in cancer research. Here we
present [Lancet](https://github.com/omicsnut/v2_lancet), an accurate and sensitive somatic variant caller, which detects
SNVs and indels by jointly analyzing reads from tumor and matched normal samples using colored de Bruijn graphs.

# Getting Started

## Tutorials

*Insert tutorial for how to do a simple run and the different options available to the user*

## Examples

*Example of running Lancet with the bam and reference files in repo*

# Guides

## Testing

*How to run tests?*

## Benchmarking

*How to benchmark. Brief explanation/reference to how the benchmark is determined*

[Benchmarking Repo](https://github.com/nygenome/benchmark/tree/master/virtual_tumor/evaluation)

<a href="https://zhubry.github.io/LancetDocumentationExample/indel_20220528.performance-plots.pdf" target="_blank">Indel
benchmark</a>

<a href="https://zhubry.github.io/LancetDocumentationExample/snv_20220528.performance-plots.pdf" target="_blank">SNV
benchmark</a>

## Scripts

*Discuss scripts that make it easier to utilize Lancet*

# Specs/Reference

## Memory Utilization

*How memory is utilized in Lancet*

<a href="https://zhubry.github.io/LancetDocumentationExample/Memory_Utilized_vs_Length_SNV.pdf" target="_blank">SNV
memory</a>

<a href="https://zhubry.github.io/LancetDocumentationExample/Memory_Utilized_vs_Length_indel.pdf" target="_blank">Indel
memory</a>

## Threads/Cores and Runtime

*How do the core hours change*

<a href="https://zhubry.github.io/LancetDocumentationExample/Core_Hours_vs_Length_SNV.pdf" target="_blank">SNV runtime (
16 threads)</a>

<a href="https://zhubry.github.io/LancetDocumentationExample/Core_Hours_vs_Length_indel.pdf" target="_blank">Indel
runtime (16 threads)</a>

## Standards/references/docs

*Add references*

# Explanation

## This is Lancet

*Explanation of how Lancet works behind the scenes*