"use strict";(self.webpackChunklancet_2_website=self.webpackChunklancet_2_website||[]).push([[468],{3905:(e,t,n)=>{n.d(t,{Zo:()=>s,kt:()=>m});var o=n(7294);function l(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);t&&(o=o.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,o)}return n}function a(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){l(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function r(e,t){if(null==e)return{};var n,o,l=function(e,t){if(null==e)return{};var n,o,l={},i=Object.keys(e);for(o=0;o<i.length;o++)n=i[o],t.indexOf(n)>=0||(l[n]=e[n]);return l}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(o=0;o<i.length;o++)n=i[o],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(l[n]=e[n])}return l}var d=o.createContext({}),p=function(e){var t=o.useContext(d),n=t;return e&&(n="function"==typeof e?e(t):a(a({},t),e)),n},s=function(e){var t=p(e.components);return o.createElement(d.Provider,{value:t},e.children)},u={inlineCode:"code",wrapper:function(e){var t=e.children;return o.createElement(o.Fragment,{},t)}},h=o.forwardRef((function(e,t){var n=e.components,l=e.mdxType,i=e.originalType,d=e.parentName,s=r(e,["components","mdxType","originalType","parentName"]),h=p(n),m=l,c=h["".concat(d,".").concat(m)]||h[m]||u[m]||i;return n?o.createElement(c,a(a({ref:t},s),{},{components:n})):o.createElement(c,a({ref:t},s))}));function m(e,t){var n=arguments,l=t&&t.mdxType;if("string"==typeof e||l){var i=n.length,a=new Array(i);a[0]=h;var r={};for(var d in t)hasOwnProperty.call(t,d)&&(r[d]=t[d]);r.originalType=e,r.mdxType="string"==typeof e?e:l,a[1]=r;for(var p=2;p<i;p++)a[p]=n[p];return o.createElement.apply(null,a)}return o.createElement.apply(null,n)}h.displayName="MDXCreateElement"},7905:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>d,contentTitle:()=>a,default:()=>u,frontMatter:()=>i,metadata:()=>r,toc:()=>p});var o=n(7462),l=(n(7294),n(3905));const i={id:"cli",title:"Command Line Options"},a=void 0,r={unversionedId:"cli",id:"cli",title:"Command Line Options",description:"pipeline",source:"@site/docs/cli.md",sourceDirName:".",slug:"/cli",permalink:"/Lancet2/docs/cli",draft:!1,editUrl:"https://github.com/facebook/docusaurus/tree/main/packages/create-docusaurus/templates/shared/docs/cli.md",tags:[],version:"current",frontMatter:{id:"cli",title:"Command Line Options"}},d={},p=[{value:"pipeline",id:"pipeline",level:2},{value:"General Options:",id:"general-options",level:2},{value:"<code>--help</code>",id:"--help",level:3},{value:"<code>--version</code>",id:"--version",level:3},{value:"<code>--verbose</code>",id:"--verbose",level:3},{value:"Required Arguments:",id:"required-arguments",level:2},{value:"<code>-t</code>, <code>--tumor</code>",id:"-t---tumor",level:3},{value:"<code>-n</code>, <code>--normal</code>",id:"-n---normal",level:3},{value:"<code>-r</code>, <code>--reference</code>",id:"-r---reference",level:3},{value:"<code>-o</code>, <code>--out-vcf</code>",id:"-o---out-vcf",level:3},{value:"Optional Arguments:",id:"optional-arguments",level:2},{value:"<code>--graphs-dir</code>",id:"--graphs-dir",level:3},{value:"Regions",id:"regions",level:3},{value:"<code>--region</code>",id:"--region",level:3},{value:"<code>-b</code>, <code>--bed-file</code>",id:"-b---bed-file",level:3},{value:"<code>-P</code>, <code>--padding</code>",id:"-p---padding",level:3},{value:"<code>-w</code>, <code>--window-size</code>",id:"-w---window-size",level:3},{value:"<code>-T</code>, <code>--num-threads</code>",id:"-t---num-threads",level:3},{value:"<code>--pct-overlap</code>",id:"--pct-overlap",level:3},{value:"Parameters",id:"parameters",level:3},{value:"<code>-T</code>, <code>--num-threads</code>",id:"-t---num-threads-1",level:3},{value:"<code>-k</code>, <code>--min-kmer-length</code>",id:"-k---min-kmer-length",level:3},{value:"<code>-K</code>, <code>--max-kmer-length</code>",id:"-k---max-kmer-length",level:3},{value:"<code>--min-trim-qual</code>",id:"--min-trim-qual",level:3},{value:"<code>-q</code>, <code>--min-base-qual</code>",id:"-q---min-base-qual",level:3},{value:"<code>-Q</code>, <code>--min-mapping-qual</code>",id:"-q---min-mapping-qual",level:3},{value:"<code>--max-rpt-mismatch</code>",id:"--max-rpt-mismatch",level:3},{value:"<code>--max-tip-length</code>",id:"--max-tip-length",level:3}],s={toc:p};function u(e){let{components:t,...n}=e;return(0,l.kt)("wrapper",(0,o.Z)({},s,n,{components:t,mdxType:"MDXLayout"}),(0,l.kt)("h2",{id:"pipeline"},"pipeline"),(0,l.kt)("p",null,"This is the subcommand that will kick off the tool. This will always follow directly after the call to the Lancet executable"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"./lancet pipeline --help\n./lancet pipeline -t /path/to/tumor.bam -n /path/to/normal.bam -r /path/to/ref.fasta -o /path/to/out.vcf\n")),(0,l.kt)("h2",{id:"general-options"},"General Options:"),(0,l.kt)("h3",{id:"--help"},(0,l.kt)("inlineCode",{parentName:"h3"},"--help")),(0,l.kt)("p",null,"Bring up the standard help output. This will give examples of options the user can change to customize a run."),(0,l.kt)("h3",{id:"--version"},(0,l.kt)("inlineCode",{parentName:"h3"},"--version")),(0,l.kt)("p",null,"Print the version information for the build"),(0,l.kt)("h3",{id:"--verbose"},(0,l.kt)("inlineCode",{parentName:"h3"},"--verbose")),(0,l.kt)("p",null,"Turns on verbose logging and will allow the user to see all log outputs"),(0,l.kt)("h2",{id:"required-arguments"},"Required Arguments:"),(0,l.kt)("p",null,"A standard run of Lancet will provide a tumor bam file, a normal bam file, the reference fasta, and an output path to place the outputted vcf file."),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"./lancet pipeline -t /path/to/tumor.bam -n /path/to/normal.bam -r /path/to/ref.fasta -o /path/to/out.vcf\n")),(0,l.kt)("h3",{id:"-t---tumor"},(0,l.kt)("inlineCode",{parentName:"h3"},"-t"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--tumor")),(0,l.kt)("p",null,"Provide the path to the tumor bam file. Index for this file should also be in the same directory"),(0,l.kt)("h3",{id:"-n---normal"},(0,l.kt)("inlineCode",{parentName:"h3"},"-n"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--normal")),(0,l.kt)("p",null,"Provide the path to the normal bam file. Index for this file should also be in the same directory"),(0,l.kt)("h3",{id:"-r---reference"},(0,l.kt)("inlineCode",{parentName:"h3"},"-r"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--reference")),(0,l.kt)("p",null,"Provide the path to the reference fasta file. Index for this file should also be in the same directory"),(0,l.kt)("h3",{id:"-o---out-vcf"},(0,l.kt)("inlineCode",{parentName:"h3"},"-o"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--out-vcf")),(0,l.kt)("p",null,"Specify where to place the outputted vcf file.`"),(0,l.kt)("h2",{id:"optional-arguments"},"Optional Arguments:"),(0,l.kt)("p",null,"These arguments allow for more fine-tuned control of the tool. If not provided, default values will be assigned"),(0,l.kt)("h3",{id:"--graphs-dir"},(0,l.kt)("inlineCode",{parentName:"h3"},"--graphs-dir")),(0,l.kt)("p",null,"This tag allows you to define the output path for dumping serialized graphs from a run. If this option is not utilized, there will be no outputted graphs."),(0,l.kt)("h3",{id:"regions"},"Regions"),(0,l.kt)("p",null,"These options will allow you to play around with what the tool looks at."),(0,l.kt)("h3",{id:"--region"},(0,l.kt)("inlineCode",{parentName:"h3"},"--region")),(0,l.kt)("p",null,"Allows the user to define what region the tool should run on. Should be of format chr:start_pos-end_pos.\nFor example, this will indicate that the tool will run on chromosome 2 between positions 33091000 and 33092000:"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"... --region 2:33091000-33092000 ...\n")),(0,l.kt)("p",null,"If no region(s) specified, the tool will default to run on the whole genome provided"),(0,l.kt)("h3",{id:"-b---bed-file"},(0,l.kt)("inlineCode",{parentName:"h3"},"-b"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--bed-file")),(0,l.kt)("p",null,"Path to bed file that will be used to define which region the tool will run on. Bed file should be a tab delimited file in which the first column\nis the chromosome, the second column is the start position, and the third column is the end position. For example, a bed file like this specifies chromosomes\n1 and 2 between positions 4 and 239410 and 892 and 1029348 respectively:"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"1   4   239410\n2   892 1029348\n")),(0,l.kt)("p",null,"If no region(s) specified, the tool will default to run on the whole genome provided"),(0,l.kt)("h3",{id:"-p---padding"},(0,l.kt)("inlineCode",{parentName:"h3"},"-P"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--padding")),(0,l.kt)("p",null,"Padding to be applied to all input regions. By default, there will be a padding of 250 bp."),(0,l.kt)("h3",{id:"-w---window-size"},(0,l.kt)("inlineCode",{parentName:"h3"},"-w"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--window-size")),(0,l.kt)("p",null,"This tag allows you to define the window size used for each microassembly task. By default, the window will be 600 bp."),(0,l.kt)("h3",{id:"-t---num-threads"},(0,l.kt)("inlineCode",{parentName:"h3"},"-T"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--num-threads")),(0,l.kt)("p",null,"Allows the user to specify the number of cores to be used for the tool."),(0,l.kt)("h3",{id:"--pct-overlap"},(0,l.kt)("inlineCode",{parentName:"h3"},"--pct-overlap")),(0,l.kt)("p",null,"Allows the user to define how much overlap there should be between windows. If not specified, the tool will default to 50% overlap"),(0,l.kt)("h3",{id:"parameters"},"Parameters"),(0,l.kt)("p",null,"These options allow you to define certain parameters for how the tool performs its variant calling"),(0,l.kt)("h3",{id:"-t---num-threads-1"},(0,l.kt)("inlineCode",{parentName:"h3"},"-T"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--num-threads")),(0,l.kt)("p",null,"This allows you to define how many threads are used by the tool. If not specified, the tool will default to running with 1 thread."),(0,l.kt)("h3",{id:"-k---min-kmer-length"},(0,l.kt)("inlineCode",{parentName:"h3"},"-k"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--min-kmer-length")),(0,l.kt)("p",null,"This allows you to define the minimum length kmers should be for graph nodes. If no length specified, the min kmer length defaults to 11 bp."),(0,l.kt)("h3",{id:"-k---max-kmer-length"},(0,l.kt)("inlineCode",{parentName:"h3"},"-K"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--max-kmer-length")),(0,l.kt)("p",null,"This allows you to define the maximum length kmers should be for graph nodes. If no length specified, the max kmer length defaults to 101 bp."),(0,l.kt)("h3",{id:"--min-trim-qual"},(0,l.kt)("inlineCode",{parentName:"h3"},"--min-trim-qual")),(0,l.kt)("p",null,"This allows you to define the minimum base quality for trimming 5' and 3' read bases. If this option is not used, the minimum base quality will be defaulted to 10."),(0,l.kt)("h3",{id:"-q---min-base-qual"},(0,l.kt)("inlineCode",{parentName:"h3"},"-q"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--min-base-qual")),(0,l.kt)("p",null,"This allows you to define the minimum base quality for SNV calling. If this option is not used, the minimum base quality will be defaulted to 17."),(0,l.kt)("h3",{id:"-q---min-mapping-qual"},(0,l.kt)("inlineCode",{parentName:"h3"},"-Q"),", ",(0,l.kt)("inlineCode",{parentName:"h3"},"--min-mapping-qual")),(0,l.kt)("p",null,"This allows you to define the minimum mapping quality required to use a read. By default, this value will be set to 15."),(0,l.kt)("h3",{id:"--max-rpt-mismatch"},(0,l.kt)("inlineCode",{parentName:"h3"},"--max-rpt-mismatch")),(0,l.kt)("p",null,"This allows you to define the maximum number of mismatches used to detect approximate repeats. By default, the maximum number of mismatches is set to 2."),(0,l.kt)("h3",{id:"--max-tip-length"},(0,l.kt)("inlineCode",{parentName:"h3"},"--max-tip-length")),(0,l.kt)("p",null,"This will define the maximum tip length allowed in the genereated graphs. By default, the maximum tip length is set to 11."),(0,l.kt)("h3",{id:""}))}u.isMDXComponent=!0}}]);