<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=author content="Rajeeva Lochan Musunuri"><link rel="canonical" href="https://nygenome.github.io/Lancet2/guides/custom_visualization.html"><link rel=prev href=targeted_analysis.html><link rel=next href=../reference.html><link rel=icon href=../assets/favicon.ico><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.6.14"><title>Custom Visualization - Lancet2</title><link rel=stylesheet href=../assets/stylesheets/main.342714a4.min.css><link rel=stylesheet href=../assets/stylesheets/palette.06af60db.min.css><script src="../assets/external/unpkg.com/iframe-worker/shim.js"></script><link rel="stylesheet" href="../assets/external/fonts.googleapis.com/css.49ea35f2.css"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#custom-visualization class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../index.html title=Lancet2 class="md-header__button md-logo" aria-label=Lancet2 data-md-component=logo> <img src=../assets/logo_nygenome.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Lancet2 </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Custom Visualization </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media=(prefers-color-scheme) data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=black data-md-color-accent=indigo aria-label="Switch to system preference" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to system preference" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href="https://github.com/nygenome/Lancet2" title="Go to repository" class="md-source" data-md-component="source"> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> nygenome/Lancet2 </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../index.html class=md-tabs__link> Getting Started </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=scoring_somatic_variants.html class=md-tabs__link> Guides </a> </li> <li class=md-tabs__item> <a href=../reference.html class=md-tabs__link> Reference </a> </li> <li class=md-tabs__item> <a href=../publications.html class=md-tabs__link> Publications </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../index.html title=Lancet2 class="md-nav__button md-logo" aria-label=Lancet2 data-md-component=logo> <img src=../assets/logo_nygenome.svg alt=logo> </a> Lancet2 </label> <div class=md-nav__source> <a href="https://github.com/nygenome/Lancet2" title="Go to repository" class="md-source" data-md-component="source"> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> nygenome/Lancet2 </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../index.html class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex> <span class=md-ellipsis> Guides </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Guides </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=scoring_somatic_variants.html class=md-nav__link> <span class=md-ellipsis> Scoring Somatic Variants </span> </a> </li> <li class=md-nav__item> <a href=wgs_analysis.html class=md-nav__link> <span class=md-ellipsis> Whole Genome Analysis </span> </a> </li> <li class=md-nav__item> <a href=targeted_analysis.html class=md-nav__link> <span class=md-ellipsis> Targeted Analysis </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Custom Visualization </span> <span class="md-nav__icon md-icon"></span> </label> <a href=custom_visualization.html class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Custom Visualization </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#inspecting-dot-formatted-assembly-graphs class=md-nav__link> <span class=md-ellipsis> Inspecting DOT formatted assembly graphs </span> </a> <nav class=md-nav aria-label="Inspecting DOT formatted assembly graphs"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#low_cov_removal1 class=md-nav__link> <span class=md-ellipsis> low_cov_removal1 </span> </a> </li> <li class=md-nav__item> <a href=#found_ref_anchors class=md-nav__link> <span class=md-ellipsis> found_ref_anchors </span> </a> </li> <li class=md-nav__item> <a href=#compression1 class=md-nav__link> <span class=md-ellipsis> compression1 </span> </a> </li> <li class=md-nav__item> <a href=#low_cov_removal2 class=md-nav__link> <span class=md-ellipsis> low_cov_removal2 </span> </a> </li> <li class=md-nav__item> <a href=#compression2 class=md-nav__link> <span class=md-ellipsis> compression2 </span> </a> </li> <li class=md-nav__item> <a href=#short_tip_removal class=md-nav__link> <span class=md-ellipsis> short_tip_removal </span> </a> </li> <li class=md-nav__item> <a href=#fully_pruned class=md-nav__link> <span class=md-ellipsis> fully_pruned </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#inspecting-gfa-formatted-sequence-graphs class=md-nav__link> <span class=md-ellipsis> Inspecting GFA formatted sequence graphs </span> </a> <nav class=md-nav aria-label="Inspecting GFA formatted sequence graphs"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bandageng class=md-nav__link> <span class=md-ellipsis> BandageNG </span> </a> </li> <li class=md-nav__item> <a href=#sequence-tube-map class=md-nav__link> <span class=md-ellipsis> Sequence Tube Map </span> </a> <nav class=md-nav aria-label="Sequence Tube Map"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#demo-sequence-tube-map-server class=md-nav__link> <span class=md-ellipsis> Demo Sequence Tube Map server </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../reference.html class=md-nav__link> <span class=md-ellipsis> Reference </span> </a> </li> <li class=md-nav__item> <a href=../publications.html class=md-nav__link> <span class=md-ellipsis> Publications </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#inspecting-dot-formatted-assembly-graphs class=md-nav__link> <span class=md-ellipsis> Inspecting DOT formatted assembly graphs </span> </a> <nav class=md-nav aria-label="Inspecting DOT formatted assembly graphs"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#low_cov_removal1 class=md-nav__link> <span class=md-ellipsis> low_cov_removal1 </span> </a> </li> <li class=md-nav__item> <a href=#found_ref_anchors class=md-nav__link> <span class=md-ellipsis> found_ref_anchors </span> </a> </li> <li class=md-nav__item> <a href=#compression1 class=md-nav__link> <span class=md-ellipsis> compression1 </span> </a> </li> <li class=md-nav__item> <a href=#low_cov_removal2 class=md-nav__link> <span class=md-ellipsis> low_cov_removal2 </span> </a> </li> <li class=md-nav__item> <a href=#compression2 class=md-nav__link> <span class=md-ellipsis> compression2 </span> </a> </li> <li class=md-nav__item> <a href=#short_tip_removal class=md-nav__link> <span class=md-ellipsis> short_tip_removal </span> </a> </li> <li class=md-nav__item> <a href=#fully_pruned class=md-nav__link> <span class=md-ellipsis> fully_pruned </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#inspecting-gfa-formatted-sequence-graphs class=md-nav__link> <span class=md-ellipsis> Inspecting GFA formatted sequence graphs </span> </a> <nav class=md-nav aria-label="Inspecting GFA formatted sequence graphs"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bandageng class=md-nav__link> <span class=md-ellipsis> BandageNG </span> </a> </li> <li class=md-nav__item> <a href=#sequence-tube-map class=md-nav__link> <span class=md-ellipsis> Sequence Tube Map </span> </a> <nav class=md-nav aria-label="Sequence Tube Map"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#demo-sequence-tube-map-server class=md-nav__link> <span class=md-ellipsis> Demo Sequence Tube Map server </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=custom-visualization>Custom Visualization<a class=headerlink href=#custom-visualization title="Permanent link">&para;</a></h1> <p>For each window being processed, Lancet2 can optionally serialize two types of graphs that are generated during the variant calling process. This can be enabled by passing a non-existing directory to the <code>--graphs-dir</code> flag to write per window serialized graphs in the directory.</p> <ol> <li> <p>the kmer-based colored de-bruijn overlap graph after the cleaning/pruning steps of the assembly process serialized as <code>DOT</code> formatted <a href="https://www.graphviz.org/">graphviz</a> files. One <code>DOT</code> file is generated for every strongly connected component for each window.</p> </li> <li> <p>the sequence graph generated from the multiple sequence alignment of all the assembled contigs from the de-bruijn graph assembly serialized as <code>GFA</code> formatted files.</p> </li> </ol> <pre><code class=language-bash>Lancet2 pipeline --reference ref.fasta \
    --tumor tumor.bam --normal normal.bam \
    --region &quot;chr22&quot; --graphs-dir output-graphs
</code></pre> <p>The above command will create two directories named <code>dbg_graph</code> and <code>poa_graph</code> in the provided <code>output-graphs</code> directory.</p> <ul> <li> <p>In the <code>dbg_graph</code> directory, you will find the <code>DOT</code> files generated for each window – <code>dbg__${CHROM}_${START}_${END}__fully_pruned__k${KMER_SIZE}__comp${COMP_ID}.dot</code></p> </li> <li> <p>In the <code>poa_graph</code> directory, you will find the <code>GFA</code> files generated for each window – <code>msa__${CHROM}_${START}_${END}__c${COMP_ID}.gfa</code> The <code>poa_graph</code> also contains a <code>FASTA</code> file with the raw multiple sequence alignment output.</p> </li> </ul> <div class="admonition info"> <p class=admonition-title>Info</p> <p>If you are interested to further debug the de-bruijn graph assembly process, you can generate DOT files for the all the intermediate steps before the <code>fully_pruned</code> graph is generated, by re-compiling the Lancet2 binary in <code>Debug</code> mode using <code>-DCMAKE_BUILD_TYPE=Debug</code> instead of <code>-DCMAKE_BUILD_TYPE=Release</code> in the cmake step.</p> <p>The intermediate states that can be inspected in their order of generation are as follows – <code>low_cov_removal1</code>, <code>found_ref_anchors</code>, <code>compression1</code>, <code>low_cov_removal2</code>, <code>compression2</code>, <code>short_tip_removal</code>.</p> </div> <h2 id=inspecting-dot-formatted-assembly-graphs>Inspecting <code>DOT</code> formatted assembly graphs<a class=headerlink href=#inspecting-dot-formatted-assembly-graphs title="Permanent link">&para;</a></h2> <p>The <code>DOT</code> files can be rendered in the pdf format using the dot utility available in the <a href="https://www.graphviz.org/">graphviz</a> visualization software package. The <code>DOT</code> file can also be exported to various other <a href="https://graphviz.org/docs/outputs/">output formats</a> with the dot utility tool.</p> <pre><code class=language-bash>dot -Tpdf -o example_file.pdf example_file.dot
</code></pre> <p>The above command will create a example_file.pdf file that shows the de-bruijn assembly graph.</p> <p>Shown below are the renderings of the de-bruijn assembly graph through all of it's intermediate states (can only be visualized in <code>Debug</code> mode Lancet2 binary. ideal for small regions.).</p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>Only the <code>fully_pruned</code> graph will be serialized when the standard <code>Release</code> build is used. The remaining 6 other states can only be visualized in the <code>Debug</code> build of Lancet2. The <code>Debug</code> build is much slower than the standard <code>Release</code> build and is only recommended to be used on small regions for inspection.</p> </div> <h4 id=low_cov_removal1><code>low_cov_removal1</code><a class=headerlink href=#low_cov_removal1 title="Permanent link">&para;</a></h4> <p><img align=left, alt=low_cov_removal1 loading=lazy src=../assets/01_dbg__chr1_38506673_38507173__low_cov_removal1__k31__comp0.png></p> <h4 id=found_ref_anchors><code>found_ref_anchors</code><a class=headerlink href=#found_ref_anchors title="Permanent link">&para;</a></h4> <p><img align=left, alt=found_ref_anchors loading=lazy src=../assets/02_dbg__chr1_38506673_38507173__found_ref_anchors__k31__comp1.png></p> <h4 id=compression1><code>compression1</code><a class=headerlink href=#compression1 title="Permanent link">&para;</a></h4> <p><img align=left, alt=compression1 loading=lazy src=../assets/03_dbg__chr1_38506673_38507173__compression1__k31__comp1.png></p> <h4 id=low_cov_removal2><code>low_cov_removal2</code><a class=headerlink href=#low_cov_removal2 title="Permanent link">&para;</a></h4> <p><img align=left, alt=low_cov_removal2 loading=lazy src=../assets/04_dbg__chr1_38506673_38507173__low_cov_removal2__k31__comp1.png></p> <h4 id=compression2><code>compression2</code><a class=headerlink href=#compression2 title="Permanent link">&para;</a></h4> <p><img align=left, alt=compression2 loading=lazy src=../assets/05_dbg__chr1_38506673_38507173__compression2__k31__comp1.png></p> <h4 id=short_tip_removal><code>short_tip_removal</code><a class=headerlink href=#short_tip_removal title="Permanent link">&para;</a></h4> <p><img align=left, alt=short_tip_removal loading=lazy src=../assets/06_dbg__chr1_38506673_38507173__short_tip_removal__k31__comp1.png></p> <h4 id=fully_pruned><code>fully_pruned</code><a class=headerlink href=#fully_pruned title="Permanent link">&para;</a></h4> <p><img align=left, alt=fully_pruned loading=lazy src=../assets/07_dbg__chr1_38506673_38507173__fully_pruned__k31__comp1.png></p> <h2 id=inspecting-gfa-formatted-sequence-graphs>Inspecting <code>GFA</code> formatted sequence graphs<a class=headerlink href=#inspecting-gfa-formatted-sequence-graphs title="Permanent link">&para;</a></h2> <p>The generated <a href="https://github.com/GFA-spec/GFA-spec"><code>GFA</code></a> files can be visualized in <a href="https://github.com/rrwick/Bandage">Bandage</a> or <a href="https://github.com/asl/BandageNG">BandageNG</a>.</p> <p>The GFA file generated by Lancet2 is chopped i.e the nodes are in their decomposed form. They need to be unchopped using the <a href="https://github.com/vgteam/vg/releases"><code>vg toolkit</code></a>, before they can be visualized or used in other downstream processes. Below is a command to unchop the Lancet2 GFA graph.</p> <pre><code class=language-bash>vg mod --unchop ${INPUT_GFA} &gt; ${OUTPUT_GFA}
</code></pre> <h4 id=bandageng>BandageNG<a class=headerlink href=#bandageng title="Permanent link">&para;</a></h4> <p><img align=left, alt=vg_unchopped_gfa_bandageNG loading=lazy src=../assets/08_vg_unchopped_gfa_bandageNG.png></p> <p>The image above shows a screenshot of an unchopped Lancet2 sequence graph in <code>GFA</code> format visualized using the BandageNG toolkit.</p> <h4 id=sequence-tube-map>Sequence Tube Map<a class=headerlink href=#sequence-tube-map title="Permanent link">&para;</a></h4> <p>Lancet2 addresses the longstanding challenge of visualizing variants along with supporting reads from multiple samples in graph space. Users can easily load the unchopped <code>GFA</code> formatted graphs into the Sequence Tube Map environment allowing somatic variants' visualization along with their supporting read alignments from multiple samples.</p> <p>The following steps describe the workflow to visualize Lancet2 variants of interest in graph space using the Sequence Tube Map framework.</p> <ol> <li> <p>Install prerequisite tools – Lancet2 (<a href="https://github.com/nygenome/Lancet2">https://github.com/nygenome/Lancet2</a>), samtools, bcftools, vg version 1.59.0, and jq, and ensure that they are available as commands that can be executed in the environment PATH.</p> </li> <li> <p>Install Sequence Tube Map (<a href="https://github.com/vgteam/sequenceTubeMap">https://github.com/vgteam/sequenceTubeMap</a>) version <a href="https://github.com/vgteam/sequenceTubeMap/tree/0452ecb82d057372e359a9b456d789336e5ab8a1">0452ecb82d057372e359a9b456d789336e5ab8a1</a>.</p> </li> <li> <p>Use the Lancet2’s prep_stm_viz.sh script to run Lancet2 on a small set of variants of interest that need to be visualized in Sequence Tube Map. The script will run Lancet2 using the --graph-dir flag to generate GFA formatted sequence graphs for each variant of interest. Local VG graphs and indices required to load the sample reads along with the Lancet2 graph are then constructed enabling simplified use with “custom” data option in Sequence Tube Map interface.</p> </li> <li> <p>After running the Sequence Tube Map Server as detailed in the Tube Map Readme, set “Data” to “custom” and “BED file” to “index.bed”. Pick a “Region”(or variant) of interest and hit “Go” to visualize.</p> </li> </ol> <h5 id=demo-sequence-tube-map-server>Demo Sequence Tube Map server<a class=headerlink href=#demo-sequence-tube-map-server title="Permanent link">&para;</a></h5> <p>Sequence Tube Map demo server containing somatic variants from COLO829 tumor/normal pair sample can be found at this link – <a href="https://shorturl.at/JQZsy">https://shorturl.at/JQZsy</a></p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-datetime" title="June 17, 2025 00:30:08 UTC">June 17, 2025 00:30:08</span> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener"> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <div class=md-progress data-md-component=progress role=progressbar></div> <script id=__config type=application/json>{"base": "..", "features": ["content.code.copy", "navigation.top", "navigation.tabs", "navigation.instant", "navigation.instant.prefetch", "navigation.instant.progress"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <!-- Add scripts that need to run before here --> <script defer src="../assets/external/cloud.umami.is/script.js" data-website-id="f52133d2-5a6a-437e-bb81-66244bf29b69"></script> <script src=../assets/javascripts/bundle.13a4f30d.min.js></script> <!-- Add scripts that need to run afterwards here --> </body> </html>